Fix module definition jumping in Swift interface files

aelam opened the pull request:
# Fix module definition jumping for submodules and Swift interface files

## Problem Description

### 1. Submodule Import Navigation Issues
- **Foundation.NSArray**: Clicking on `NSArray` in `import Foundation.NSArray` fails to jump to definition
- **Darwin Submodules**: Clicking on `sysdir` in `import Darwin.sysdir` fails to jump to definition  
- **Root Cause**: SourceKit-LSP cannot properly handle module names containing dots (module.submodule)

### 2. Swift Interface File Navigation Problems
- **Generated Interface Navigation**: Command+Click navigation functionality is broken within generated `.swiftinterface` files
- **Multiple Interface Caching**: Repeatedly opening the same module interface generates multiple duplicate files
- **Language Service Mapping**: Interface files cannot be properly mapped to corresponding language services

## Solution

### 1. Enhanced Module Symbol Processing

**File**: `Sources/SourceKitLSP/SourceKitLSPServer.swift`

#### Module Name Parsing Logic
```swift
// Enhanced module symbol processing
if symbol.kind == .module {
  let moduleName: String
  let groupName: String?
  
  if let systemModule = symbol.systemModule {
    // Prefer using systemModule information
    moduleName = systemModule.moduleName
    groupName = systemModule.groupName
  } else if let name = symbol.name {
    // Handle module names containing dots like "Darwin.sysdir"
    if name.contains(".") {
      let components = name.split(separator: ".", maxSplits: 1)
      moduleName = String(components[0])
      groupName = components.count == 2 ? String(components[1]) : nil
    } else {
      moduleName = name
      groupName = nil
    }
  }
}
```

#### System Module Navigation
```swift
// System module navigation handling
if symbol.isSystem ?? false, let systemModule = symbol.systemModule {
  return try await self.definitionInInterface(
    moduleName: systemModule.moduleName,
    groupName: systemModule.groupName,
    symbolUSR: symbol.usr,
    originatorUri: uri,
    languageService: languageService
  )
}
```

### 2. Swift Interface File Support

#### Generated Interface Manager Enhancement
**File**: `Sources/SourceKitLSP/Swift/GeneratedInterfaceManager.swift`

```swift
// Improved interface file caching
private var interfaceDocuments: [InterfaceKey: GeneratedInterfaceDocument] = [:]

// Avoid duplicate generation of interface files for the same module
struct InterfaceKey: Hashable {
  let moduleName: String
  let groupName: String?
  let buildSettings: String
}
```

#### Language Service Mapping Fix  
**File**: `Sources/SourceKitLSP/Workspace.swift`

```swift
// Fix language service mapping for interface files
func setDocumentService(_ service: DocumentService, for uri: DocumentURI) {
  let key = buildSettingsFile ?? uri
  documentService[key] = service
}
```

### 3. Interface File Recognition
**File**: `Sources/SourceKitLSP/Swift/OpenInterface.swift`

```swift
// Recognize .swiftinterface file extensions
private static let swiftInterfaceExtensions: Set<String> = ["swiftinterface"]

var isSwiftInterface: Bool {
  swiftInterfaceExtensions.contains(pathExtension.lowercased())
}
```

## Test Cases

### 1. Foundation Submodules
✅ `import Foundation.NSArray` → Correctly jumps to NSArray interface  
✅ `import Foundation.NSURL` → Correctly jumps to NSURL interface

### 2. Darwin Submodules  
✅ `import Darwin.sysdir` → Correctly jumps to sysdir interface  
✅ `import Darwin.uuid` → Correctly jumps to uuid interface  
✅ `import Darwin.POSIX` → Correctly jumps to POSIX interface

### 3. Swift Interface Navigation
✅ Command+Click navigation works properly within generated `.swiftinterface` files  
✅ Avoids duplicate generation of interface files for the same module  
✅ Symbols in interface files can correctly jump to definitions

## Technical Details

### Key Improvements

1. **Module Name Parsing**: Properly handles `module.submodule` format import statements
2. **System Module Support**: Uses `systemModule` information for accurate module navigation  
3. **Interface File Caching**: Avoids duplicate generation, improves performance
4. **Language Service Mapping**: Ensures interface files can be properly mapped to language services
5. **File Type Recognition**: Correctly identifies `.swiftinterface` file types

### Backward Compatibility

- ✅ Maintains normal functionality of existing import statements
- ✅ Does not affect navigation behavior of non-system modules
- ✅ Compatible with existing interface generation mechanisms

## Impact

### Feature Improvements
1. **Submodule Navigation**: Full support for Foundation and Darwin submodule navigation
2. **Interface File Navigation**: Swift interface file navigation functionality restored
3. **Performance**: Reduced duplicate interface file generation, improved cache efficiency
4. **Stability**: Fixed interface file language service mapping issues

## Related Issues

- Fix Foundation.NSArray and other submodule import jumping issues
- Fix Darwin.sysdir and other system submodule navigation
- Resolve navigation failures within generated Swift interface files
- Optimize interface file caching mechanism to avoid duplicate generation

## Testing Recommendations

It is recommended to test in the following scenarios:
1. Various Foundation submodule import jumping
2. Darwin system submodule import jumping  
3. Navigation within generated Swift interface files
4. Caching behavior when opening the same module interface multiple times
5. Ensure normal module imports are not affected

![symbols](https://github.com/user-attachments/assets/aa911290-5353-4644-a1e6-627d1f263aea)

ahoppen commented on Sources/SourceKitLSP/Swift/GeneratedInterfaceManager.swift:
```diff
@@ -92,7 +92,13 @@ actor GeneratedInterfaceManager {
-      guard let cachedIndex = openInterfaces.firstIndex(where: { $0.url == document }) else {
+      // Cache by module name and group name, not the full document data
+      // This allows reuse across different buildSettingsFrom URIs
+      guard let cachedIndex = openInterfaces.firstIndex(where: { 
```
I don’t think this is correct. Depending on the build settings used to retrieve the generated interface, the interface might look differently. I can’t come up with a great real-world example right now where the contents differ but a canonical example would be that we can’t load the generated interface if we are missing an include path to one of the module’s dependencies but we can generate the interface if all build settings are present.

But if we continue passing along the same `buildSettingsFrom` file, I don’t think we need this change. Or am I missing something?

aelam replied:
you are right

but about same `buildSettingsFrom` file  please check here https://github.com/swiftlang/sourcekit-lsp/pull/2233#discussion_r2269006407

ahoppen commented on Sources/SourceKitLSP/Swift/ReferenceDocumentURL.swift:
```diff
@@ -165,13 +165,14 @@ extension DocumentURI {
       return referenceDocument.buildSettingsFile
     }
     return self
+  }  /// Convert sourcekit-lsp:// URIs to actual file system paths when possible.
```
Missing newlines before `///`?

ahoppen commented on Sources/SourceKitLSP/SourceKitLSPServer.swift:
```diff
@@ -553,7 +553,7 @@ package actor SourceKitLSPServer {
 
     let toolchain = await workspace.buildServerManager.toolchain(
-      for: uri,
+      for: uri.buildSettingsFile,
```
Is this change necessary? I just checked and it looks like `toolchain(for:in:language:)` doesn’t even use `uri` right now.

ahoppen commented on Sources/SourceKitLSP/SourceKitLSPServer.swift:
```diff
@@ -2051,9 +2065,36 @@ extension SourceKitLSPServer {
     languageService: LanguageService
   ) async throws -> Location {
+    // Check if we're already in the target interface with the same module/group/symbol
+    if let referenceDoc = try? ReferenceDocumentURL(from: originatorUri),
+       case .generatedInterface(let interfaceData) = referenceDoc,
```
I think you can slightly shorten this

```suggestion
    if case .generatedInterface(let interfaceData) = try? ReferenceDocumentURL(from: originatorUri),
```

ahoppen commented on Sources/SourceKitLSP/SourceKitLSPServer.swift:
```diff
@@ -2051,9 +2065,36 @@ extension SourceKitLSPServer {
+    // Check if we're already in the target interface with the same module/group/symbol
+    if let referenceDoc = try? ReferenceDocumentURL(from: originatorUri),
+       case .generatedInterface(let interfaceData) = referenceDoc,
+       interfaceData.moduleName == moduleName && interfaceData.groupName == groupName {
```
Similarly to my other comments about the build settings, we need to check here that we are using build settings from the same original file. But I think as long as we pass along the same build settings file, this shouldn’t cause any issues.

aelam commented on Sources/SourceKitLSP/SourceKitLSPServer.swift:
```diff
@@ -2051,9 +2065,36 @@ extension SourceKitLSPServer {
+      
+      // If we have a specific symbol USR, try to find its position in the current interface
+      if let symbolUSR = symbolUSR,
+         let swiftLanguageService = languageService as? SwiftLanguageService {
```
adding a group might be a good idea 

i imagine there could be different versions of swiftinterface which are used by different modules/files

i think the challenge is how to determine it in the  beginning instead of a uuid
then the cache logic could be simplified

aelam commented on Sources/SourceKitLSP/Swift/GeneratedInterfaceDocumentURLData.swift:
```diff
@@ -67,7 +67,14 @@ package struct GeneratedInterfaceDocumentURLData: Hashable, ReferenceURLData {
     self.groupName = groupName
     self.sourcekitdDocumentName = sourcekitdDocumentName
-    self.buildSettingsFrom = primaryFile
+    if let referenceDocumentURL = try? ReferenceDocumentURL(from: primaryFile),
+       let fileURL = referenceDocumentURL.buildSettingsFile.fileURL {
+      self.buildSettingsFrom = DocumentURI(fileURL)
+    } else if let fileURL = primaryFile.fileURL {
+      self.buildSettingsFrom = DocumentURI(fileURL)
+    } else {
+      self.buildSettingsFrom = primaryFile
+    }
```
@ahoppen 

make sure buildSettingsFrom can be only an actual file otherwise

it could be 
sourcekit-lsp://xxx?buildSettingsFrom=A
A=sourcekit-lsp://xxx?buildSettingsFrom=B

NOTE: this happens if we enable navigation inside swiftinterface in this PR

ahoppen commented on Tests/SourceKitLSPTests/SwiftInterfaceTests.swift:
```diff
@@ -319,6 +306,137 @@ final class SwiftInterfaceTests: XCTestCase {
+    let interfaceUri = try XCTUnwrap(stringDefinition?.locations?.only?.uri)
+    let interfaceContents = try await testClient.send(GetReferenceDocumentRequest(uri: interfaceUri))
+    // Open the interface document
+    testClient.send(
+      DidOpenTextDocumentNotification(
+        textDocument: TextDocumentItem(
+          uri: interfaceUri,
+          language: .swift,
+          version: 0,
+          text: interfaceContents.content
+        )
+      )
+    )
```
I think you can simplify this slightly to 


```suggestion
    testClient.openDocument(interfaceContents.content, interfaceUri, language: .swift)
```

ahoppen commented on Tests/SourceKitLSPTests/SwiftInterfaceTests.swift:
```diff
@@ -319,6 +306,137 @@ final class SwiftInterfaceTests: XCTestCase {
+    )
+
+    // Find a symbol within the interface (e.g., "init" method in String)
+    guard let initRange = interfaceContents.content.range(of: "public init()") else {
+      XCTFail("Could not find 'public init()' in String interface")
+      return
+    }
```
Can be simplified to the following, I think.


```suggestion
    let initRange = try XCTUnwrap(interfaceContents.content.range(of: "public init()"))
```

ahoppen commented on Tests/SourceKitLSPTests/SwiftInterfaceTests.swift:
```diff
@@ -319,6 +306,137 @@ final class SwiftInterfaceTests: XCTestCase {
+    )
+
+    // Find a symbol within the interface (e.g., "init" method in String)
+    guard let initRange = interfaceContents.content.range(of: "public init()") else {
```
Could we pick a function that’s unique within the Swift interface (eg. `public func withUTF8Buffer`) to make this test less likely to change if new initializers get introduced.

ahoppen commented on Tests/SourceKitLSPTests/SwiftInterfaceTests.swift:
```diff
@@ -319,6 +306,137 @@ final class SwiftInterfaceTests: XCTestCase {
+    )
+
+    // The definition should either:
+    // 1. Return a position within the same interface (internal navigation)
+    // 2. Return the same position if it's already at the definition
+    // 3. Return nil if no definition is available
```
The test should deterministically return the same results, in which case only one of these checks should be true. Could we remove the other options?

ahoppen reviewed (commented):
Thanks for your continued work on this PR and adding the tests, @aelam. I left some comments, mostly on the tests. Also, addressing the comments in SourceKitLSPServer.swift would be important to make the next review round meaningful.

aelam commented on Tests/SourceKitLSPTests/SwiftInterfaceTests.swift:
```diff
@@ -319,6 +306,137 @@ final class SwiftInterfaceTests: XCTestCase {
+        GetReferenceDocumentRequest.method: .dictionary(["supported": .bool(true)])
+      ])
+    )
+    let uri = DocumentURI(for: .swiftinterface)
```
I assumed user is already in interface file
since it makes no difference I can remove it

aelam commented:
>  Also, addressing the comments in SourceKitLSPServer.swift 

I went through all but maybe I missed any you may want to highlight? 

Sorry about too many unrelated changes, I just reverted a lot of unneccessary change and `resovle conversation`  to make it clean

ahoppen commented on Sources/SourceKitLSP/SourceKitLSPServer.swift:
```diff
@@ -2051,9 +2065,36 @@ extension SourceKitLSPServer {
     languageService: LanguageService
   ) async throws -> Location {
+    // Check if we're already in the target interface with the same module/group/symbol
+    if case .generatedInterface(let interfaceData) = try? ReferenceDocumentURL(from: originatorUri),
```
This comment still applies here: https://github.com/swiftlang/sourcekit-lsp/pull/2233/#discussion_r2267852544

ahoppen commented on Tests/SourceKitLSPTests/SwiftInterfaceTests.swift:
```diff
@@ -319,6 +323,64 @@ final class SwiftInterfaceTests: XCTestCase {
+        GetReferenceDocumentRequest.method: .dictionary(["supported": .bool(true)])
+      ])
+    )
+    let uri = DocumentURI(for: .swiftinterface)
+
+    let positions = testClient.openDocument(
+      """
+      import 1️⃣Foundation
+      """,
+      uri: uri,
+      language: .swift
+    )
```
This comment still applies here: https://github.com/swiftlang/sourcekit-lsp/pull/2233/#discussion_r2272261588. This should be able to become


```suggestion
    let uri = DocumentURI(for: .swift)

    let positions = testClient.openDocument(
      """
      import 1️⃣Foundation
      """,
      uri: uri
    )
```

It doesn’t make sense of the user to start in a `.swiftinterface` file they found somewhere on disk because we don’t have a build system that would be able to provide build settings for that `.swiftinterface` file – we only have build settings for swift interface files that were generated by SourceKit-LSP.

aelam commented on Sources/SourceKitLSP/Swift/OpenInterface.swift:
```diff
@@ -21,10 +21,18 @@ extension SwiftLanguageService {
   ) async throws -> GeneratedInterfaceDetails? {
+    // Generate a deterministic document name based on module name and group name
+    // This ensures we reuse the same interface for the same module/group combination
+    let documentName = if let groupName {
```
please check the latest change. 
it's like `module.group.buildSettingFileHash`
but I think it's still not perfect. 

In my understanding A.swift and B.swift would generate different hash. then it would be different Foundation.hash1.swiftinterface and Foundation.hash2.swiftinterface

I tried to use the workspace.rootUri hash, it can fix the above issue
but the workspace is the same for sourcekit-lsp and the `.build/checkout/IndexStoreDB/IndexStoreDB.swift`. is it the expected behavior?

```swift
// Get workspace to use its rootUri for build context identification
let workspace = await self.workspaceForDocument(uri: originatorUri)
let buildContextId = workspace?.rootUri?.stringValue ?? originatorUri.buildSettingsFile.stringValue
let buildContextHash = "\(abs(buildContextId.hashValue))"
```

ahoppen commented on Tests/SourceKitLSPTests/SwiftInterfaceTests.swift:
```diff
@@ -319,6 +319,63 @@ final class SwiftInterfaceTests: XCTestCase {
+      DefinitionRequest(textDocument: TextDocumentIdentifier(uri), position: positions["1️⃣"])
+    )
+    let foundationLocation = try XCTUnwrap(foundationDefinition?.locations?.only)
+        XCTAssertEqual(foundationLocation.uri.scheme, "sourcekit-lsp")  
```
Could you run swift format to format your changes?

ahoppen reviewed (commented):
Just one formatting issue left, otherwise LGTM.

ahoppen commented:
@swift-ci Please test Windows

aelam commented on Tests/SourceKitLSPTests/SwiftInterfaceTests.swift:
```diff
@@ -319,6 +319,63 @@ final class SwiftInterfaceTests: XCTestCase {
+  }
+
+  func testFoundationSubmoduleNavigation() async throws {
+    let testClient = try await TestSourceKitLSPClient(
```
Thanks for very detailed code every time  😄

ahoppen commented:
@swift-ci Please test Windows

ahoppen commented:
Congrats on your first contribution to SourceKit-LSP, @aelam.
